Wrappers/Python/cil/optimisation/algorithms/FISTA.py

```diff
@@ -69,51 +140,187 @@ def __init__(self, initial=None, f=None, g=ZeroFunction(), **kwargs):
             else:
                 raise ValueError('{} received both initial and the deprecated x_init parameter. It is not clear which one we should use.'\
                     .format(self.__class__.__name__))
-        
-        if initial is not None and f is not None:
-            self.set_up(initial=initial, f=f, g=g)
 
-    def set_up(self, initial, f, g=ZeroFunction()):
-        '''initialisation of the algorithm
+        self._step_size = None
+
+        # set up FISTA
+        self.set_up(initial=initial, f=f, g=g, step_size=step_size, **kwargs)
+
+    def set_up(self, initial, f, g, step_size, **kwargs):
+        """ Set up of the algorithm
+        """
+
+        self.initial = initial
+        self.f = f
+        self.g = g
+
+        # set step_size
+        self._set_step_size(step_size=step_size)
 
-        :param initial: Initial guess ( Default initial = 0)
-        :param f: Differentiable function
-        :param g: Convex function with " simple " proximal operator'''
+        # check convergence criterion for FISTA is satisfied
+        if kwargs.get('check_convergence_criterion', True):
+            self._check_convergence_criterion()
 
         print("{} setting up".format(self.__class__.__name__, ))
-        
+
         self.y = initial.copy()
         self.x_old = initial.copy()
         self.x = initial.copy()
         self.u = initial.copy()
```

Comment:
The update rule in the ISTA algorithm class does not use self.y or self.u (which are image data objects). Arguably FISTA also does not need to store self.u. Rather, it seems that all the operations on self.u can be done as operations on self.x (if optimising the memory usage is a concern)

Reply:
@paskino @epapoutsellis  I think this is a good reason to switch this to have `ISTA` as the parent class. I thought there was just an extra couple of unused scalars stored, but storing unused  copies of data is a whole different issue.

Reply:
> Arguably FISTA also does not need to store self.u

@zeljkozeljko 

That's a great spot! I think this is probably true. but it relies on `self.g.proximal` being safe to use in-place which it may not be. I think there's potential there but it needs a bit of thought. I'll open up an issue.

Reply:
For these gradient ( +proximal) methods algorithms, the inheritance should be `GD` > `ISTA` > `FISTA`. The main change would be on the `update` method of `ISTA` and `FISTA` and many adaptive strategies of the step size of `GD` can be used by `ISTA` and `FISTA`, e.g., [armijo rule](https://github.com/TomographicImaging/CIL/blob/master/Wrappers/Python/cil/optimisation/algorithms/GD.py#L109).